promising/logging_utils.py

```
@@ -0,0 +1,119 @@
+        return f"  {label}: [{status}] {ctx}"
+
+    def _unsettled_children_snapshot(self, parent: "PromisingContext") -> tuple["PromisingContext", ...]:
+        num_retries = 10
+
+        for attempt in range(num_retries):
+            try:
+                return tuple(parent._unsettled_children)
+            except RuntimeError:
+                if attempt < num_retries - 1:
+                    self.logger.log(
+                        self.level, f"Retrying unsettled-children snapshot (attempt {attempt + 1}/{num_retries})"
```

Ruff (0.15.10)

[warning] Logging statement uses f-string

(G004)